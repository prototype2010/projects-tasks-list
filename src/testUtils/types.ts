declare global {
  interface HTMLElement {
    value: any;
  }
}

export {};
