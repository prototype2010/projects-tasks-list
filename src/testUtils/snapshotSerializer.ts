import pretty from 'pretty';
import htmlParser from 'html-parser';

const snapshotSanitizers = [
  (snapshot: string) =>
    snapshot.replace(/<[a-z]* +data\-remove\-from\-snapshot*.*>(.*)<\/[a-z]*>/gi, function (match) {
      return match.replace(/>(.*)</, function () {
        return `>** REMOVED FROM SNAPSHOT **<`;
      });
    }),
];

const makePrettyHTML = (content: string) => {
  const sanitized = htmlParser.sanitize(content, {
    elements: ['script', 'img', 'svg', 'style', 'head', 'noscript'],
    comments: true,
    attributes: ['style', 'class', 'role', 'tabindex', 'aria-haspopup', 'aria-labelledby', 'id'],
    docTypes: true,
  });

  const sanitizedVariableData = snapshotSanitizers.reduce((cumulative, current) => {
    return current(cumulative);
  }, sanitized);

  return pretty(sanitizedVariableData, { ocd: true });
};

export default {
  serialize(val: any) {
    return makePrettyHTML(val.container.innerHTML);
  },

  test(val: any) {
    return !!val?.container?.innerHTML;
  },
};
