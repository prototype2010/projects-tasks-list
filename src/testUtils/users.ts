export type testCredentials = { email: string; password: string };

export const generalUserCredentials: testCredentials = {
  email: 'testemail@mail.ru',
  password: '1111111111111111',
};
