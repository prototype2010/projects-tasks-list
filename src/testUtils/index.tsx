import React from 'react';
import { Provider } from 'react-redux';
import { Router } from 'react-router';
import { createMemoryHistory } from 'history';
import { render as rtlRender, RenderResult, act } from '@testing-library/react';
import { waitFor, fireEvent } from '@testing-library/react';

import { ThemeProvider } from 'styled-components/macro';
import MaterialThemeProvider from '@material-ui/styles/ThemeProvider';
import { StylesProvider } from '@material-ui/core/styles';
import App from '../App';

import theme, { classNamesGenerator, GlobalStyles, materialTheme } from '../theme';
import { createStore } from '../store/store';
import { testCredentials } from './users';

function render(ui = App as any) {
  const history = createMemoryHistory({ initialEntries: ['/'] });

  function Wrapper({ children: Children }: any) {
    return (
      <Router history={history}>
        <Provider store={createStore()}>
          <StylesProvider generateClassName={classNamesGenerator}>
            <MaterialThemeProvider theme={materialTheme}>
              <ThemeProvider theme={theme}>
                <GlobalStyles />
                <Children />
              </ThemeProvider>
            </MaterialThemeProvider>
          </StylesProvider>
        </Provider>
      </Router>
    );
  }
  return {
    ...rtlRender(ui, { wrapper: Wrapper }),
    history,
  };
}

// re-export everything
export * from '@testing-library/react';

const clickElement = async (selector: any) => {
  await waitFor(() => selector);

  fireEvent.click(selector, { target: { button: 0 } });
};

const navigateViaSideMenu = async (app: RenderResult, path: Array<string>) => {
  for await (const menuPath of path) {
    await clickElement(app.getByTestId(menuPath));
  }
};

const makeInput = (selector: any, value: string) => {
  fireEvent.change(selector, { target: { value } });
};

const waitForAppToLoad = async ({ container }: RenderResult) => {
  return waitFor(() => null, { container });
};

const loginAs = async (app: RenderResult, { email, password }: testCredentials) => {
  const { getByText, getByLabelText, getByTestId } = app;

  await waitForAppToLoad(app);

  makeInput(getByLabelText('Email'), email);
  makeInput(getByLabelText('Password'), password);

  await clickElement(getByTestId('sign-in'));

  await waitForAppToLoad(app);

  await waitFor(() => getByText('SIMPLE TODO LISTS'));
};

export {
  render, // override render method
  clickElement,
  loginAs,
  makeInput,
  navigateViaSideMenu,
  waitForAppToLoad,
};
