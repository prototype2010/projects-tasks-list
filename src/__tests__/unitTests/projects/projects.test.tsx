import { render, cleanup, loginAs, waitForAppToLoad, clickElement, makeInput } from '../../../testUtils';
import { RenderResult, waitFor } from '@testing-library/react';
import { generalUserCredentials } from '../../../testUtils/users';

let app: RenderResult;

beforeEach(async () => {
  app = render();
  await loginAs(app, generalUserCredentials);
  await waitForAppToLoad(app);
});

afterEach(cleanup);

describe('Tasks', () => {
  it('Can be created', async () => {
    const { getByTestId, getByPlaceholderText } = app;

    const taskName = 'GREAT_TASK_#';
    await makeInput(getByPlaceholderText('Start typing here to create a new task...'), taskName);

    await clickElement(getByTestId('add-task-for-project-94'));

    await waitFor(() => getByTestId(`delete-task-${taskName}`));

    expect(app).toMatchSnapshot();
  });

  it('Can be deleted', async () => {
    const { getByTestId } = app;

    await clickElement(getByTestId('delete-project-94'));
    await clickElement(getByTestId('proceed-delete'));

    expect(app).toMatchSnapshot();
  });

  it('Can be edited', async () => {
    const { getByTestId, getByText, getByLabelText } = app;

    const newProjectName = 'NEW AWESOME NAME';

    await clickElement(getByTestId('edit-project-94'));
    await makeInput(getByLabelText('Project name'), newProjectName);
    await clickElement(getByTestId('proceed'));

    await waitFor(() => getByText(newProjectName));

    expect(app).toMatchSnapshot();
  });
});
