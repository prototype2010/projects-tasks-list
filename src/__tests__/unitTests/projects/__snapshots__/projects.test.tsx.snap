// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks Can be created 1`] = `
<div>
  <header>
    <h1>SIMPLE TODO LISTS</h1>
    <h5>FROM RUBY GARAGE</h5>
  </header>
  <div>
    <div>
      <div><span><span>GREAT PROJECT</span></span><span><div></div><div></div></span></div>
      <div>
        <div><input data-testid="project-input-94" placeholder="Start typing here to create a new task..." value><button data-testid="add-task-for-project-94">Add Task</button></div>
      </div>
      <div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-1"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>111111111</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-2"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>222222222</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-62"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>GREAT_TASK_#</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
      </div>
    </div>
  </div><button data-testid="create-project"> Add TODO List</button>
  <div aria-hidden="true">
    <div></div>
  </div>
</div>
`;

exports[`Tasks Can be deleted 1`] = `
<div>
  <header>
    <h1>SIMPLE TODO LISTS</h1>
    <h5>FROM RUBY GARAGE</h5>
  </header>
  <div>
    <div>
      <div><span><span>GREAT PROJECT</span></span><span><div></div><div></div></span></div>
      <div>
        <div><input data-testid="project-input-94" placeholder="Start typing here to create a new task..." value><button data-testid="add-task-for-project-94">Add Task</button></div>
      </div>
      <div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-1"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>111111111</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-2"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>222222222</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
      </div>
    </div>
  </div><button data-testid="create-project"> Add TODO List</button>
  <div aria-hidden="true">
    <div></div>
  </div>
</div>
`;

exports[`Tasks Can be edited 1`] = `
<div>
  <header>
    <h1>SIMPLE TODO LISTS</h1>
    <h5>FROM RUBY GARAGE</h5>
  </header>
  <div>
    <div>
      <div><span><span>NEW AWESOME NAME</span></span><span><div></div><div></div></span></div>
      <div>
        <div><input data-testid="project-input-94" placeholder="Start typing here to create a new task..." value><button data-testid="add-task-for-project-94">Add Task</button></div>
      </div>
      <div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-1"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>111111111</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
        <div>
          <div><span aria-disabled="false" data-testid="task-checkbox-2"><span><input type="checkbox" data-indeterminate="false" aria-label="secondary checkbox" value></span><span></span></span></div>
          <div>222222222</div>
          <div>
            <div>
              <div></div>
            </div>
            <div></div>
            <div></div>
          </div>
        </div>
      </div>
    </div>
  </div><button data-testid="create-project"> Add TODO List</button>
  <div aria-hidden="true">
    <div></div>
  </div>
</div>
`;
